cou/utils/openstack.py

```diff
@@ -13,118 +13,138 @@
-    },
-    "ovn-central": {"representative_workload_pkg": "ovn-common", "origin_setting": "source"},
-    "ceph-mon": {"representative_workload_pkg": "ceph-common", "origin_setting": "source"},
-    "placement": {
-        "representative_workload_pkg": "placement-common",
-        "origin_setting": "openstack-origin",
-    },
-    "nova-cloud-controller": {
-        "representative_workload_pkg": "nova-common",
-        "origin_setting": "openstack-origin",
-    },
+    "ceph": ["ceph-mon", "ceph-fs", "ceph-radosgw", "ceph-osd"],
+    "swift": ["swift-proxy", "swift-storage"],
+    "nova": ["nova-cloud-controller", "nova-compute"],
+    "ovn": ["ovn-dedicated-chassis", "ovn-central"],
+    "neutron": ["neutron-api", "neutron-gateway"],
+    "manila": ["manila-ganesha"],
+    "horizon": ["openstack-dashboard"],
 }
 
 
-def get_os_code_info(package, pkg_version) -> str:
-    """Determine OpenStack codename that corresponds to package version.
-
-    :param package: Package name
-    :type package: string
-    :param pkg_version: Package version
-    :type pkg_version: string
-    :returns: Codename for package
-    :rtype: string
-    """
-    # Remove epoch if it exists
-    if ":" in pkg_version:
-        pkg_version = pkg_version.split(":")[1:][0]
-    if "swift" in package:
-        # Fully x.y.z match for swift versions
-        match = re.match(r"^(\d+)\.(\d+)\.(\d+)", pkg_version)
-    else:
-        # x.y match only for 20XX.X
-        # and ignore patch level for other packages
-        match = re.match(r"^(\d+)\.(\d+)", pkg_version)
-
-    if match:
-        vers = match.group(0)
-    # Generate a major version number for newer semantic
-    # versions of openstack projects
-    major_vers = vers.split(".")[0]
-    if package in PACKAGE_CODENAMES and major_vers in PACKAGE_CODENAMES[package]:
-        return PACKAGE_CODENAMES[package][major_vers]
-    else:
-        # < Liberty co-ordinated project versions
-        if "swift" in package:
-            return get_swift_codename(vers)
-        elif "ovn" in package:
-            return get_ovn_codename(vers)
-        else:
-            return OPENSTACK_CODENAMES[vers]
-
-
-# Codename and package versions
-def get_swift_codename(version):
-    """Determine OpenStack codename that corresponds to swift version.
-
-    :param version: Version of Swift
-    :type version: string
-    :returns: Codename for swift
-    :rtype: string
-    """
-    return _get_special_codename(version, SWIFT_CODENAMES)
-
-
-def get_ovn_codename(version):
-    """Determine OpenStack codename that corresponds to OVN version.
-
-    :param version: Version of OVN
-    :type version: string
-    :returns: Codename for OVN
-    :rtype: string
-    """
-    return _get_special_codename(version, OVN_CODENAMES)
-
-
-def _get_special_codename(version, codenames):
-    found = [k for k, v in six.iteritems(codenames) if version in v]
-    return found[0]
+@dataclass(frozen=True)
+class VersionRange:
```

Comment:
I tried to use like this:
```python3
from collections.abc import Container

@dataclass(frozen=True)
class VersionRange(Container):
    lower: str
    upper: str
    ......
```

`mypy` complains with this:
```
lint: commands[2]> mypy --install-types --non-interactive .
cou/utils/openstack.py:45: error: Argument 1 of "__contains__" is incompatible with supertype "Container"; supertype defines the argument type as "object"  [override]
Found 1 error in 1 file (checked 11 source files)
cou/utils/openstack.py:45: note: This violates the Liskov substitution principle
cou/utils/openstack.py:45: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
lint: exit 1 (0.44 seconds) /home/cocenza/Documents/canonical/charmed-openstack-upgrader-1> mypy --install-types --non-interactive . pid=323823
  lint: FAIL code 1 (7.78=setup[0.03]+cmd[0.22,7.09,0.44] seconds)
  evaluation failed :( (7.81 seconds)
make: *** [Makefile:28: lint] Error 1
```
I think it's because `Container` is an abstract class, than it violates the Liskov substitution principle

Reply:
I gave a deeper look and if we change to something like this it will work:

```
@dataclass(frozen=True)
class VersionRange(Container):
    lower: str
    upper: str

    def __contains__(self, version: Any) -> bool:
```

The problem was that in the abstract implementation `x` is more generic than the `str`. changing to `Any` or using `# type: ignore[override]` will solve this mypy problem, but I think it does not worth inheriting it. It would be the same inheriting `Hashable` if we want to implement a custom `__hash__` method.